dask/dataframe/io/csv.py

```diff
@@ -17,17 +17,96 @@
     CategoricalDtype,
 )
 
+from ...base import tokenize
+
 # this import checks for the importability of fsspec
 from ...bytes import read_bytes, open_file, open_files
+from ..core import new_dd_object, DataFrame
+from ...core import flatten
 from ...delayed import delayed
-from ...utils import asciitable, parse_bytes
+from ...utils import asciitable, parse_bytes, ensure_dict
 from ..utils import clear_known_categories
 from .io import from_delayed
 
 import fsspec.implementations.local
 from fsspec.compression import compr
 
 
+class CSVSubgraph(Mapping):
+    def __init__(
+        self,
+        name,
+        reader,
+        blocks,
+        is_first,
+        head,
+        header,
+        kwargs,
+        dtypes,
+        columns,
+        collection,
+        enforce,
+        path,
+    ):
+        self.name = name
+        self.reader = reader
+        self.blocks = blocks
+        self.is_first = is_first
+        self.head = head if collection else None  # example pandas DF for metadata
+        self.header = header #prepend to all blocks
+        self.kwargs = kwargs
+        self.dtypes = dtypes
+        self.columns = columns
+        self.collection = collection
+        self.enforce = enforce
+        self.colname, self.paths = path or (None, None)
+
+    def __getitem__(self, key):
+        try:
+            name, i = key
+        except ValueError:
+            # too many / few values to unpack
+            raise KeyError(key) from None
+
+        if name != self.name:
+            raise KeyError(key)
+
+        if i < 0 or i >= len(self.blocks):
+            raise KeyError(key)
+
+        block = self.blocks[i]
+
+        if self.paths is not None:
+            path_info = (self.colname, self.paths[i], self.paths)
+        else:
+            path_info = None
+
+        write_header = False
+        rest_kwargs = self.kwargs.copy()
+        if not self.is_first[i]:
+            write_header = True
+            rest_kwargs.pop("skiprows", None)
+
+        return (pandas_read_text,
```

Comment:
Seems to be a few failures related to the changed result type of `text_block_to_pandas`. I view this as an internal function so I think it's OK to just update the tests.